gdc_rnaseq_tools/augment_star_counts.py

```diff
@@ -0,0 +1,228 @@
+            + 'annotation tables. Possible file truncation.'
+        )
+
+    # save table of mis-aligned read stats also reported by star
+    mdf = cdf.iloc[0:4].copy()
+
+    return df, mdf
+
+
+def calc_tpm(df, excol='unstranded', lencol='total_exon_length'):
+    '''
+    Transcripts Per Million
+
+    Calculate the TPM values for all genes in DataFrame df
+
+    TPM = RPK x 1e6 / M
+
+        RPK = C X 1e3 / L
+        C : count of fragments aligned to this gene
+        L : sum of exon lengths in gene where overlapping exons are merged
+            otherwise known as union exon length
+        M : sum over all genes of RPK values
+    '''
+    # RPK - reads per thousand bp of transcript length
+    rpk = df[excol] * 1e3 / df[lencol]
+    # sum of RPK signal
+    M = rpk.sum()
+    # TPM - transcripts per million
+    tpm = rpk * 1e6 / M
+    return tpm
+
+
+def calc_fpkm(df, excol='unstranded', lencol='total_exon_length', typecol='gene_type'):
+    '''
+    Fragments Per Kilobases (of transcript) and Millions (of fragments)
+
+    Calculate the FPKM values for all genes in DataFrame df
+
+    FPKM = (C × 1e9)/(NL)
+
+        C : count of fragments aligned to this gene
+        N : total fragment count to protein-coding genes
+        L : sum of exon lengths in gene where overlapping exons are merged
+    '''
+    # select protein coding genes
+    sel = df[typecol] == 'protein_coding'
+    # get sum of counts in protein coding genes
+    N = df.loc[sel][excol].sum()
+    # calculate fpkm
+    fpkm = df[excol] * 1e9 / (N * df[lencol])
+
+    return fpkm
+
+
+def calc_fpkm_uq(
+    df,
+    excol='unstranded',
+    lencol='total_exon_length',
+    typecol='gene_type',
+    chrcol='Chromosome',
+):
+    '''
+    Upper Quartile normalized FPKM
+
+    FPKM-UQ = (C × 1e9)/(UGL)
+
+        C : count of fragments aligned to this gene
+        U : upper quartile of fragment counts to autosomal protein-coding genes with count > 0
+        G : number of protein-coding genes on autosomes
+        L : sum of exon lengths in gene where overlapping exons are merged
+    '''
+    # selections for U and G
+    sel_prot = df[typecol] == 'protein_coding'
+    sel_autosomes = ~df[chrcol].isin(['chrX', 'chrY', 'chrM'])
+    sel_nonzero = df[excol] > 0
+    sel_U = sel_prot & sel_autosomes & sel_nonzero
+    sel_G = sel_prot & sel_autosomes
+
+    # Calculate U and G
+    U = np.quantile(df[sel_U][excol], 0.75)
+    G = len(df[sel_G])
+
+    # Calculate FPKM-UQ
+    fpkm_uq = df[excol] * 1e9 / (U * G * df[lencol])
+    return fpkm_uq
+
+
+def save_result(df, outfile, pragma_line=''):
+    '''
+    Write output table as TSV
+    '''
+
+    out_cols = [
+        'gene_id',
+        'gene_name',
+        'gene_type',
+        'stranded_first',
+        'stranded_second',
+        'unstranded',
+        'tpm_unstranded',
+        'fpkm_unstranded',
+        'fpkm_uq_unstranded',
+    ]
+
+    with open(outfile, 'w') as out:
+        print(pragma_line, file=out)
```

Comment:
```suggestion
        out.write(pragma_line + "\n")
```
I would prefer the write function rather than printing out to a file.